azafea/event_processors/endless/metrics/machine.py

```diff
@@ -67,3 +93,27 @@ def upsert_machine_live(dbsession: DbSession, machine_id: str) -> None:
                                       set_={'live': True})
 
     dbsession.connection().execute(stmt)
+
+
+def upsert_machine_location(dbsession: DbSession, machine_id: str,
+                            info: Union[Dict[str, Any], List[Any]]) -> None:
+    """Update the relevant Machine record with information from a LocationLabel event.
+
+    Although the info is guaranteed to be a `Dict[str, str]` when received from
+    the client, the database column is of type JSONB, so the
+    `LocationLabel.info` field has a more general type. As a result, the
+    function parameter also has a more general type, and this runtime type
+    check is needed to satisfy mypy.
+
+    """
+    if isinstance(info, dict):
+        location_columns = [column.name for column in Machine.__table__.columns]
+        values = {
+            f'location_{key}': value for key, value in info.items()
+            if f'location_{key}' in location_columns}
+        stmt = insert(Machine.__table__).values(machine_id=machine_id, location=info, **values)
+        stmt = stmt.on_conflict_do_update(constraint='uq_metrics_machine_machine_id', set_=values)
+
+        dbsession.connection().execute(stmt)
+    else:  # pragma: no cover
+        log.warning('Data received for machine location is not a dict: {%s}', info)
```

Comment:
To me the `{}`s are strongly associated with dictionaries so perhaps just:

```suggestion
        log.warning('Data received for machine location is not a dict: %r', info)
```

Reply:
Sure.